notes/plutus-core/cek-budgeting-profiling/CekProfiling.md

```diff
@@ -0,0 +1,381 @@
+wasn't visible in the primetest example (closer examination of the profiling
+data suggests it took something like 0.6% of the execution time).  This is startup
+overhead: it's called when the initial AST is annotated with the memory usage of nodes
+(and I didn't remove this in the versions with `spendBudget` removed).
+
+The picture for `Restricting` mode is very similar to that for `Counting`.
+
+### marlowe/zerocoupon/1
+
+![marlowe/zerocoupon/1: Counting mode](./zc1-counting.svg)
+
+![marlowe/zerocoupon/1: Restricting mode](./zc1-restricting.svg)
+
+The graphs for the zerocoupon example are a little strange, with some
+data floating in midair.  This is for CBOR decoding functions called via
+typeclasses; I'm not sure why it's not attached to the main graph.
+
+The situation is broadly similar to the crowdfunding example.  The main
+CEK function `evaluateCek` takes 73.6% of the total evaluation time
+and `spendBudget` takes 40.7%  (29.7% + 1%), so it takes up 55%
+of the main CEK evaluation function (again, this would suggest a 2x slowdown,
+but we saw a 3.39x slowdown in the benchmark table).
+
+
+## Conclusions on budgeting
+
+It seems clear that budgeting is taking a lot of time in the CEK machine,
+particularly where hashmaps are used.  We should probably try to simplify things
+as much as possible, possibly using some other data structure and reducing the
+use of lenses, monads, and typeclasses.
+
+
+### No budgeting
+
+The graphs below show what happens when we have no budgeting, and so give some
+idea of what's taking time in the basic CEK machine.
+
+![marlowe/zerocoupon/1: no budgeting](./zc1-no-budgeting.svg)
+
+![primetest/20digits: no budgeting](./prime20-no-budgeting.svg)
+
+![crowdfunding/1: no budgeting](./cf1-no-budgeting.svg)
+
+For the two validation examples much of the time is consumed by startup
+overhead: in both the crowdfunding example and the zerocoupon example only 25%
+of the total execution time is spent in `evaluateCek`; in contrast, the much
+longer-running primetest example spends 94% of its time there.  We should
+definitely try to make deserialisation as efficient as possible.  I've used CBOR
+here, but fortunately Radu's benchmarks suggest that `flat` deserialisation is
+generally faster, taking about 75-95% of the time required to deserialise CBOR.
+
+Quite a lot of time seems to be spent looking up names in environments and
+inserting new names (about 80% for crowdfunding, 60% for zerocoupon, 25% for
+primetest).  Using de Bruijn levels could help with this: we already know that
+de Bruijn _indices_ help to reduce script sizes, and presumably the same would
+apply for levels.  However, if we can compute directly with de Bruijn levels it
+might be possible to look things up more rapidly: since we know in advance how
+deep recursion will be I think we could pre-allocate an array of mutable
+variables to store the terms referred to by the levels, and this would allow
```

Comment:
Yeah, I'm not sure about that. I thought we could have a sort of stack of bindings, but maybe that'd require us to start re-indexing variables inside terms, which we want to avoid.  We should probably try to implement something, then any problems would become more apparent.

Reply:
Sure, but then you either have
- A stack of mutable arrays (which you then can't pre-allocate); or
- A mutable array of stacks (in which case you probably aren't benefiting that much from the mutability; or
- You use machine frames to restore old variable bindings when you leave (in which case you're using the GHC stack and you're anyway maybe making something more like a stack machine)

Reply:
That's not exactly what I meant.  I had a vague idea that you could have one big long mutable array and effectively use it as a stack, advancing further along every time you add a new binding and retreating when that binding goes out of scope (perhaps exposing an earlier binding for the same variable).  I didn't think it through though, and I think I'd have to start implementing something before I could see if that made sense.